examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,39 +18,152 @@
-        chip::CharSpan()
-        // TODO: Convert responses to Optional<chip::app::DataModel::List<const
-        // chip::app::Clusters::Messages::Structs::MessageResponseOptionStruct::Type>> message.responses = responses;
-    };
-
-    mMessages.push_back(message);
+    std::vector<char *> needToFree;
+    std::vector<MessageResponseOption *> optionToFree;
+
+    uint8_t * messageIdBuffer = nullptr;
+    char * messageTextBuffer  = new char[messageText.size()];
+    VerifyOrExit(messageTextBuffer != nullptr,
+                 ChipLogProgress(Controller, "HandlePresentMessagesRequest messageTextBuffer alloc failed"));
+    needToFree.push_back(messageTextBuffer);
+    memcpy(messageTextBuffer, messageText.data(), messageText.size());
+
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest message:%s size:%lu", messageTextBuffer, messageText.size());
+
+    messageIdBuffer = new uint8_t[messageId.size()];
+    VerifyOrExit(messageIdBuffer != nullptr,
+                 ChipLogProgress(Controller, "HandlePresentMessagesRequest messageIdBuffer alloc failed"));
+    memcpy(messageIdBuffer, messageId.data(), messageId.size());
+
+    if (responses.HasValue())
+    {
+        size_t size    = 0;
+        CHIP_ERROR err = responses.Value().ComputeSize(&size);
+        VerifyOrExit(err == CHIP_NO_ERROR, ChipLogProgress(Controller, "HandlePresentMessagesRequest size check failed"));
+
+        MessageResponseOption * optionArray = new MessageResponseOption[size];
+        VerifyOrExit(optionArray != nullptr,
+                     ChipLogProgress(Controller, "HandlePresentMessagesRequest MessageResponseOption alloc failed"));
+        optionToFree.push_back(optionArray);
+
+        int counter = 0;
+        auto iter   = responses.Value().begin();
+        while (iter.Next())
+        {
+            auto & response = iter.GetValue();
+
+            ChipLogProgress(Controller, "HandlePresentMessagesRequest option id:%u", response.messageResponseID.ValueOr(0));
+
+            if (response.messageResponseID.HasValue())
+            {
+                optionArray[counter].messageResponseID = Optional<uint32_t>(response.messageResponseID.ValueOr(0));
+            }
+
+            if (response.label.HasValue())
+            {
+                char * labelBuffer = new char[response.label.Value().size()];
+                VerifyOrExit(labelBuffer != nullptr,
+                             ChipLogProgress(Controller, "HandlePresentMessagesRequest label alloc failed"));
+                needToFree.push_back(labelBuffer);
+                memcpy(labelBuffer, response.label.Value().data(), response.label.Value().size());
+
+                ChipLogProgress(Controller, "HandlePresentMessagesRequest option label:%s size:%lu", labelBuffer,
+                                response.label.Value().size());
+
+                optionArray[counter].label = Optional<CharSpan>(CharSpan::fromCharString(labelBuffer));
+            }
+            counter++;
+        }
+
```

Comment:
Where's the check that the iterator did not hit errors?  This code is silently allowing invalid TLV.

Reply:
Added early exit when the option is incomplete